File: 5.longest-palindromic-substring.cpp
/*
 * @lc app=leetcode id=5 lang=cpp
 *
 * [5] Longest Palindromic Substring
 */

// @lc code=start
class Solution {
public:
    string longestPalindrome(string s) {
        int start = 0;
        int len = 0;
        int n = s.size();
        for (int i = 0; i < s.size(); i++) {
            int l = i, r = i;
            while (r < n - 1 && s[l] == s[r + 1]) {
                r++;
                i++;
            }
            while (l > 0 && r < n - 1 && s[l - 1] == s[r + 1]) {
                l--;
                r++;
            }
            int m = r - l + 1;
            if (m > len) {
                len = m;
                start = l;
            }
        }

        return s.substr(start, len);
    }
};
// @lc code=end

File: 5.longest-palindromic-substring.java
/*
 * @lc app=leetcode id=5 lang=java
 *
 * [5] Longest Palindromic Substring
 */

// @lc code=start
class Solution {
    public String longestPalindrome(String s) {
        int max = 0, l = 0, r = 0, n = s.length();
        for (int i = 0; i < s.length(); i++) {
            int a = i, b = i;
            while (a >= 0 && b < n && s.charAt(a) == s.charAt(b)) {
                --a;
                ++b;
            }
            if (b - a > max) {
                max = b - a;
                l = a + 1;
                r = b - 1;
            }
            a = i;
            b = i + 1;
            while (a >= 0 && b < n && s.charAt(a) == s.charAt(b)) {
                --a;
                ++b;
            }
            if (b - a > max) {
                max = b - a;
                l = a + 1;
                r = b - 1;
            }
        }
        return s.substring(l, r + 1);
    }
}
// @lc code=end

